merkle/src/merkle.rs

```diff
@@ -248,6 +258,231 @@ impl<E: Element> Clone for MmapStore<E> {
     }
 }
 
+/// File-mapping version of `MmapStore` with the added `new_with_path` method
+/// that allows to set its path (otherwise a temporary file is used which is
+/// cleaned up after we drop this structure).
+#[derive(Debug)]
+pub struct DiskMmapStore<E: Element> {
+    // Implementing the `store` with `Arc`/`RwLock` to avoid adding lifetimes
+    // parameters to the struct (which might have larger repercussions in the
+    // definitions of the `MerkleTree` and its consumers. Also used for its
+    // coordination mechanisms, but it's not clearly defined if `MerkleTree`
+    // (and `Store`) should be thread-safe (so they might be removed later).
+    store: Arc<RwLock<Option<MmapMut>>>,
+
+    len: usize,
+    _e: PhantomData<E>,
+    file: File,
+    // We need to save the `File` in case we're creating a `tempfile()`
+    // otherwise it will get cleaned after we return from `new()`.
+
+    // We cache the `store.len()` call to avoid accessing the store when
+    // it's offloaded. Not to be confused with `len`, this saves the total
+    // size of the `store` and the other one keeps track of used `E` slots
+    // in the `DiskMmapStore`.
+    store_size: usize,
+
+    // We save the arguments of `new_with_path` to reconstruct it and reload
+    // the `MmapMut` after offload has been called.
+    path: PathBuf,
+    size: Option<usize>,
+}
+
+impl<E: Element> ops::Deref for DiskMmapStore<E> {
+    type Target = [E];
+
+    fn deref(&self) -> &Self::Target {
+        unimplemented!()
+    }
+}
+
+impl<E: Element> Store<E> for DiskMmapStore<E> {
+    #[allow(unsafe_code)]
+    fn new(size: usize) -> Self {
+        let byte_len = E::byte_len() * size;
+        let file: File = tempfile().expect("couldn't create temp file");
+        file.set_len(byte_len as u64)
+            .unwrap_or_else(|_| panic!("couldn't set len of {}", byte_len));
+
+        let mmap = unsafe { MmapMut::map_mut(&file).expect("couldn't create map_mut") };
```

Comment:
Thinking about this a bit more, there are multiple operations here that could fail, we should return an error, rather than panicing, as this will otherwise kill the rest of the operations. Doesn't have to be in this PR, but should be at least a `TODO`+ and issue